src/base/static/components/input-form/input-form.js

```diff
@@ -0,0 +1,833 @@
+
+    return attrs;
+  }
+}
+
+// TODO: greensboropb hook(s)
+
+class InputForm extends Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      selectedCategory: null,
+      isCategoryMenuCollapsed: false,
+      isCategoryMenuHidden: false,
+      fieldValues: {},
+      isFormSubmitting: false,
+      formValidationErrors: [],
+      isMapPositioned: false,
+      isFormVisible: false
+    };
+
+    this.geometry = null;
+    this.geometryStyle = null;
+    this.isWithCustomGeometry = false;
+    this.requiredFields = [];
+    this.richTextFields = [];
+    this.richTextImages = [];
+    this.coverImages = [];
+    this.visibleCategories = [];
+
+    this.validationExclusions = new Set([
+      constants.COMMON_FORM_ELEMENT_TYPENAME,
+      constants.MAP_DRAWING_TOOLBAR_TYPENAME,
+      constants.PUBLISH_CONTROL_TOOLBAR_TYPENAME,
+      constants.SUBMIT_FIELD_TYPENAME
+    ]);
+
+    this.onCategoryChange = this.onCategoryChange.bind(this);
+    this.onExpandCategories = this.onExpandCategories.bind(this);
+    this.onSubmit = this.onSubmit.bind(this);
+    this.onGeometryChange = this.onGeometryChange.bind(this);
+    this.onGeometryStyleChange = this.onGeometryStyleChange.bind(this);
+    this.onCheckboxFieldChange = this.onCheckboxFieldChange.bind(this);
+    this.onAttachmentFieldChange = this.onAttachmentFieldChange.bind(this);
+    this.onFieldChange = this.onFieldChange.bind(this);
+
+    this.props.map.on("dragend", () => {
+      !this.state.isMapPositioned 
+        && this.setState({ isMapPositioned: true });
+    });
+  }
+
+  resetState() {
+    const isSingleCategory = (this.visibleCategories.length === 1);
+
+    this.setState({
+      isFormVisible: false,
+      selectedCategory: isSingleCategory ? this.visibleCategories[0] : null,
+      isCategoryMenuCollapsed: false,
+      isCategoryMenuHidden: isSingleCategory,
+      fieldValues: {},
+      isFormSubmitting: false,
+      formValidationErrors: [],
+      coverImages: [],
+      isMapPositioned: false
+    });
+    this.resetCategoryData();
+  }
+
+  resetCategoryData() {
+    this.geometry = null;
+    this.geometryStyle = null;
+    this.isWithCustomGeometry = false;
+    this.requiredFields = [];
+    this.richTextFields = [];
+    this.richTextImages = [];
+    this.coverImages = [];
+  }
+
+  componentWillMount() {
+    this.visibleCategories = this.props.placeConfig.place_detail
+      .filter(config => config.includeOnForm)
+      .filter(config => {
+        return (
+          !(config.admin_only 
+            && !Util.getAdminStatus(config.dataset, config.admin_groups))
+        );
+      });
+
+    if (this.visibleCategories.length === 1) {
+      this.setActiveCategoryState(this.visibleCategories[0]);
+      this.setState({ isCategoryMenuHidden: true });
+    }
+  }
+
+  componentDidMount() {
+    this.setState({ isFormVisible: true });
+    new Spinner(Shareabouts.smallSpinnerOptions)
```

Comment:
`Spinner` here is undefined.

Reply:
`Spinner` is globally available, but I suppose a good refactor would be to make a proper Spinner component here instead.

Reply:
Ah, I see. I'm not familiar with a global Spinner class. Do you know what API this is coming from?

I see it's in the existing code, so it's reasonable to keep for now. But I would suggest using a library to leverage this in React, although it can be done in a later PR if that's easier.

Reply:
I think it's from `spin.js`, linked to from `base.hbs`. It's from the original Shareabouts code; I think I just forgot about it in this PR, but I think porting this to React now makes sense.

Reply:
Ah, ok, it's sourced and brought in as a global :\

I've had a good experience with this library: http://www.material-ui.com/#/components/circular-progress although it may take some effort to set up. Since this PR is already huge, I think it's reasonable to wait and do it in a separate PR - it's up to you.